### ISSUE TITLE
Report unnecessary casting operations 

### ISSUE THREAD

**Issue by kalessil** [kalessil/phpinspectionsea#386]
Original link: https://bitbucket.org/kalessil/phpinspectionsea/issues/97/fr-suggest-the-removal-of-a-cast-for-the
Author: @rentalhost 

---

**Example:** if you do `!$something` it'll always cast to _boolean_, then do `(bool) !$something` is useless.
This rule can basically check if the `(cast)` type is the same that next value. So it can work with a variable number of cases, for instance:

```php
(string) $instance->__toString();
(int) ( $instance + 1 ); // only if $instance is integer too
(float) ( $instance + 1 ); // only if $instance is float too
```

**Comment by b1rdex** [comment]
This should not be applied to composite types.
I mean `int|string` should not be warned for both int and string casts.

**Comment by kalessil** [comment]
That's for sure =)

**Comment by rentalhost** [comment]
Maybe you can apply the annotation directly to casting? Like `<weak>(object)</weak>`? It just an eye candy.

**Comment by kalessil** [comment]
Good idea, thanks. Will do.
